enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,801 @@
+
+### Goals
+
+* Provide a "best practices" approach to declarative management of Control Plane Machines
+* Allow users to "1-click" scale their Control Plane to large or smaller instances
+* Allow the adoption of existing Control Planes into this new mechanism
+* Provide a safe mechanism to perform the sensitive operation of scaling the Control Plane and provide adequate
+  feedback to end users about progress of the operation
+* Allow users to opt-out of Control Plane management should our mechanism not fit their needs
+* Allow users to customise rolling update strategies based on the needs of their environments
+
+### Non-Goals
+
+* Allow horizontal scaling of the Control Plane (this may be required in the future, but today is considered
+  unsupported by OpenShift)
+* Management of the etcd cluster state (the etcd operator will handle this separately)
+* Automatic adoption of existing clusters (we will provide instructions to allow users to opt-in for existing cluster)
+* Management of anything that falls out of the scope of Machine API
+
+## Proposal
+
+A new CRD `ControlPlaneMachineSet` will be introduced into OpenShift and a respective `control-plane-machine-set-
+operator` will be introduced as a new second level operator within OpenShift to perform the operations described in
+this proposal.
+
+The new CRD will define the specification for managing (creating, adopting, updating) Machines for the Control Plane.
+The operator will be responsible for ensuring the desired number of Control Plane Machines are present within the
+cluster as well as providing update mechanisms akin to those seen in a StatefulSets and Deployments to allow rollouts
+of updated configuration to the Control Plane Machines.
+
+### User Stories
+
+- As a cluster administrator of OpenShift, I would like to be able to safely and automatically vertically resize my
+  control plane as and when the demand on the control plane changes
+- As a cluster administrator of OpenShift, I would like to be able to automatically recover failed Control Plane
+  Machines
+- As a cluster administrator of OpenShift, I would like to be able to make changes to the configuration of the
+  underlying hardware of my control plane and have these changes safely applied using immutable hardware concepts
+- As a cluster administrator of OpenShift, I would like to be able to control rollout of changes to my Control Plane
+  Machines such that I can test changes with a single replica before applying the change to all replicas
+- (Future work) As a cluster administrator of OpenShift with restricted hardware capacity, I would like to be able to   
+  scale down my control plane before adding new Control Plane Machines with newer configuration, notably, my
+  environment does not have capacity to add additional Machines during updates
+
+### API Extensions
+
+We will introduce a new `ControlPlaneMachineSet` CRD to the `machine.openshift.io/v1beta1` API group. It will be based
```

Comment:
Are *ANY* of the key user stories here applicable to generic sets of nodes and normal machine sets?  It's helpful framing here to describe why this isn't a convention around MachineSets, or a more general controller type (`StatefulMachineSet`).  I *think* you address the justifications in multiple sections, but whenever you introduce a new type you need to clearly and succinctly describe why *this type* vs the other options.  Choosing something specific to Control Plane has advantages, choosing something generic but more powerful than machine set has advantages, but as a design point it needs to be unambiguous what justifies this.

Reply:
Ok.  And ultimately the design pattern for control plane machine set allows you to be opinionated, and the concrete problem you describe is indeed a control plane problem.

Will review the comments.

Reply:
(also, i like control plane machine set because we *won't* have that for hypershift, which means not having the api fundamentally describes the type of cluster as self-hosted or not)

Reply:
> And ultimately the design pattern for control plane machine set allows you to be opinionated,

I tried to articulate this in the alternative at the bottom, a list of restrictions we are adding to make this control plane specific